### DISCUSSION TITLE
Push stats instead of pulling

### DISCUSSION THREAD

**Discussion by Hutch79** [henrygd/beszel#51] [Ideas]
Hey there,

First of all I love this project!
Finally, a monitoring solution which is very quick to set up, thanks for the work so far!

Now to my suggestion.
The current setup works (as far as I understand it) that the host reaches out to the agent and pulls the stats from it.
My suggestion would be to reverse this process, so the agent pushes the stats to the Host via an API call.

This would have two benefits.
1. Different Tokens/Key's. Currently, every Agent is set up with the same SSH-Key. With an API every Agent could have a different token which decreases the risk of accidental leaking. I guess individual SSH-Keys would also be possible with the current setup.
2. Only the Host would need to be accessible via a Port. Currently, every Agent needs a port, so the Host can access it. Iv we reverse this setup, only the Host needs an open port.
The API/Push approach would also allow for custom Agent's to be created more easily.

Anyway, I'm curious where this project goes!

**Comment by henrygd** [comment]
This has been brought up previously in #16 and #28, but this seems like a better place to discuss it.

It's on the radar for the future but not something I'm planning to jump into immediately.

For those having trouble with the current setup, it may be possible to connect using services like Cloudflare tunnel, WireGuard, or Tailscale.

**Reply by Hutch79** [reply]
I understand that this is not priority number one.
First everything must be stable as is, after that reworking existing stuff can happen.

I had connecting trough VPN's or overlay networks like Tailscale or Netbird in my mind, but my networking knowledge is not good enough to implement such a setup. Also, an open port is not a dealbreaker for me personally, but for some it might be.

**Comment by TheAnachronism** [comment]
Really hope this can be done at some point. With monitoring many VMs on a hypervisor, or systems that don't have public addresses, this would be tremendous.

**Comment by tuxdotrs** [comment]
Hey @henrygd! Are you open to pull requests? If so, I'd like to work on this.

**Reply by henrygd** [reply]
Sure, you can take a stab at it if you want.

On the agent side we'll probably want a ticker that sends the stats every 60 seconds. You can get the stats via the `gatherStats` method on the agent.

On the hub side I'm thinking maybe an mTLS endpoint is the way to go as that takes care of auth without needing to implement / manage tokens. The endpoint would receive the stats with system id and add the records. 

The agent side could define the cert / key as env vars. If those env vars exist, use the ticker to push stats rather than starting the SSH server.

Then the other tricky part is that we'll need to exclude the systems that receive stats this way from the hub's current batch update that sends SSH requests for stats every 60 sec. Off the top of my head, I'm not sure of the best way to do this. Maybe just a bool db field on the system and the user can switch the mode?

Let me know if you have any better ideas.

Edit: We might also want to check with the hub first that the system is not paused before getting / sending stats.

**Comment by twiesing** [comment]
I would very much welcome this feature! I have a large number of hosts where I can open the SSH port. But not for 2-3 hosts. A ‘multi-setup’, where some go via tailscale and others directly via the SSH port, is complicated to set up. Push would be excellent.

**Comment by skurfuerst** [comment]
Hey @henrygd ,

Just wanted to chime in - we are running NATS quite heavily and it is really just fire and forget - it simply works. The hardest part are the complexities around authentication/authorization, but for this use case can be greatly simplified by using shared tokens. And if somebody wants to do more complex things, TLS Certificates or NKeys could be used.

**If useful, I can provide my help / support implementing this**; I am an experienced dev.

We are currently migrating over from Netdata and this would be really useful to us :)

All the best,
Sebastian

**Reply by henrygd** [reply]
Hi Sebastian, that's good to know, thanks!

I've only just dipped my toes in with NATS and need spend more time with it to make sure it's the right fit.

There are a few unrelated issues I need to address first, but this is the next major feature on my list and I should be able to make progress in the next few weeks.

I'd love to collaborate and will follow up here when I have something concrete. Thanks again!

**Comment by Daniel15** [comment]
There's several good reasons to pull instead of push, which is why Prometheus does it this way: https://prometheus.io/docs/introduction/faq/#why-do-you-pull-rather-than-push

> Pulling over HTTP offers a number of advantages:
>  * You can start extra monitoring instances as needed, e.g. on your laptop when developing changes.
>  * You can more easily and reliably tell if a target is down.
> * You can manually go to a target and inspect its health with a web browser.

They offer a "push gateway" for cases where you can't pull. You push to the gateway, then Prometheus pulls from the gateway.

**Reply by ruohki** [reply]
Sometimes you simply don't have the ability to pull. I don't see why adding a push mechanism is such a bad thing. You probably won't monitor 200+ serves with Beszel and have to fear ddosing yourself.

**Reply by Daniel15** [reply]
@ruohki  I'm not saying it's bad, just that there's some advantages of pulling that are lost when pushing. That goes the other way too - there's some advantages of pushing over pulling. There's no reason Beszel couldn't support both.

**Reply by skurfuerst** [reply]
Hey @gedw99 ,

You are discussing JetStream here - which is the "exactly once delivery" part of NATS, which is optional. For the beszel use case, I would NOT use this but use simple PubSub instead (at most once delivery). This will be definitely enough for the metrics use case in Beszel.

All the best,
Sebastian

**Reply by ruohki** [reply]
Oh, think we got things wrong. I want both things push and pull depending on the situation :) @Daniel15

**Comment by freedbygrace** [comment]
In looking at this conversation. NATs would be the best solution. If it is a heavier lift in the front, and is reliable, the time spent troubleshooting later will be greatly reduced. Only if those boxes are checked though. Having worked on a number of engineering projects, simple (Cutting corners) usually results in massive overhauls (and regret) later. The price for success is paid upfront and in full to reap the benefits on the back end. It really just depends on which cost you want to pay, but the price will be paid either way.

**Reply by ruohki** [reply]
I don't think introducing an extra service for this is a good thing here. Beszel aims for simplicity.

**Comment by JSmith-Aura** [comment]
Just noting that the SSH standard library for golang is extremely well developed and allows for bidirectional communication with relatively little cost. 
I am the primary maintainer of https://github.com/NHAS/reverse_ssh and I think just using an ssh client to do all of this work seems fine to me.

**Reply by JSmith-Aura** [reply]
While I think NATs looks great, Im not sure if it's currently suitable.

Currently beszel which uses pocketbase is a single application meaning that it does not and can not have tolerance to network partitions, or failure of the underlying host.
As you've explored with pocketbase they are not open to making it a highly available application framework at this stage (https://github.com/pocketbase/pocketbase/discussions/395)

Additionally, the monitoring statistics that are collected by beszel are not what I would consider "critical data". I.e if a subsection of these records are lost there is no detriment other than not having visibly of that time period. 
As an another exmaple ,if beszel dealt with authentication details, payment records, or other incredibly important data then your desire to use NATS (and probably jetstream) would be more suitable

**Reply by JSmith-Aura** [reply]
I would also say that often time in engineering that perfect is often the enemy of good enough.  

The current beszel networking architecture is less than optimal for a number of users (including myself) so improving that and then dealing with scaling issues when it becomes a problem is probably the best approach

**Reply by AndreKR** [reply]
> The data always gets there no matter what your network or topology is doing.

The forward connection as it is currently implemented would lose data when the connection is interrupted. I think the reverse connection should behave the same.

If some store-and-forward buffering gets implemented, I feel like it should work the same for agent-to-server connections as for server-to-agent connections.

**Reply by henrygd** [reply]
I generally agree with all of this.

I've been playing around a bit with NATS. It's wonderful and I plan to use it for a lot of stuff outside of Beszel in the future. But we may not need its capabilities for what we're trying to do here.

If we can accomplish this with SSH in a way that doesn't require multiple steps to set up keys for the agent, then I think it makes perfect sense.

Apologies if I seem to be all over the place on this. I don't have as much experience with network design as you all clearly do. I'm trying to learn but haven't had a lot of time recently to do that.

On a related note, I've been thinking about adding an option for agents to keep their own database and stream data on demand that we wouldn't necessarily want to keep in hub's database. This could also allow us to fill in gaps when the connection is interrupted.

**Reply by JSmith-Aura** [reply]
I like that approach. Its very similar to what elastic-agent does which I think is a fairly mature logging project. 
The only design considerations that needs to be thought of around that is making sure you dont swam the beszel service if it goes offline for some period of time. 

Also on a side note. 

Making fairly good headway on adding this to beszel. https://github.com/JSmith-Aura/beszel

Ignore the 8 migrations Im going to clean up things before I make a PR but getting there haha. Just got to make the connection acceptance dialog, and a setting for allowing registrations when presented with an API key.

**Reply by henrygd** [reply]
@JSmith-Aura Thank you! I think this is a good approach. Please @ me if you have any questions or if I can help in any way. You can also email me: hank at henrygd.me.

**Reply by NHAS** [reply]
Just letting you know that Im switching to developing it on my person account rather than my work account haha.

**Reply by NHAS** [reply]
I've now created the first version of this (albeit poorly tested and a tad messy in some spots) here:

https://github.com/henrygd/beszel/pull/579

**Reply by ruohki** [reply]
Why do we want to open an ssh connection back to the server. Why not simply a rest request with a token or something? Connection via ssh could compromise the beszel host when a agent system gets compromised, giving an attacker potential access to all the other servers.

**Reply by NHAS** [reply]
You are confusing ssh the protocol and openssh the server software. 

SSH is like TLS. Golang has an incredibly mature ssh library that you can define anything you like with, it doesnt give you remote code execution and it cant compromise other agents or systems. 

Please go and read the code.

**Reply by AndreKR** [reply]
I agree that Go's SSH library is great and that SSH is probably a viable way. I do think HTTPs would be a *better* way though. For SSH you would have to open/forward another port and take care of host keys and such. With TLS pretty much every load balancer/reverse proxy/firewall speaks it and keys are taken care of by ACME. Wherever you deploy the Beszel server, chances are that you already have a load balancer there where you would just have to add another subdomain.

**Reply by NHAS** [reply]
I wholeheartedly agree that HTTPs might be better in this case. 

There are two caveats I can see in this case however:
- Sometimes it is quite good to separate your management interface and your "work" interface. 
  As it makes it harder for people to mess up and expose their instance if for instance they set a weak password. 
- SSH being a natively multiplexed protocol makes it very easy to add new features by adding new "channels" that you can then pass other data or functionality over

Aside from that I'd also say that the changes I've made make it quite easy to just implement an HTTPs route as there is now a concept within the beszel server of a "client" system rather than a "server" system. I.e one that connects to beszel rather than the alternative.

**Reply by ruohki** [reply]
I just dont understand why we try to reinvent the wheel and overcomplicate things. Gzip that payload and call it a day. After all this is just an option to use when your beszel instance cannot reach into the network of the to be monitored instance? I mean Pocketbase has an api :D

**Reply by NHAS** [reply]
You're more than welcome to make your own pull request or work off of what I've already done. 

Currently beszel uses ssh and I thought due to the fact it would require minimal client changes both in interface and code to implement it as an ssh client that it would be the best option. 

Also, this feature is rather important in most larger networks you dont want to write a firewall rule to allow your monitoring instance to hit each new client, or worst of all allowing blanket outbound to your hosts. 
Its better to enable things to connect to your monitoring instance. 

I would also say that an API route is not going to give you the host liveness where SSH does, so if you do want to implement an API path for this, you'll most likely have to use websockets. 

I think it's probably up to @henrygd to weigh in here. As I definitely think there is merit to just using HTTP

**Reply by henrygd** [reply]
Sorry, I'm off the computer today but I'll follow up tomorrow after I've looked more at the PR.

If HTTP is more convenient then I'm happy to go with community opinion. Maybe I'll make a poll. I'm sure we can adapt a lot of what @NHAS did in any case.

I would ask that nobody start on another PR for this right now as I don't want to waste anyone's time.

**Reply by NHAS** [reply]
No stress, this definitely isnt an urgent matter.

**Reply by ruohki** [reply]
Just an example from the real world. I host beszel on a VM on the web. I'd like to also monitor some things on my local network. Without a VPN, this is not possible right now. I really don't want to set up multiple instances.

**Comment by skurfuerst** [comment]
Hey @henrygd ,

I have previously advocated for NATS (and still think it might be a good fit), but I was thinking whether the different options *really* make a difference. 

IMHO, it totally makes sense to have a protocol which has encryption and authentication built in. Other than that, any standard protocol currently on the table (SSH as it is right now, NATS, HTTPS) has some specific pros and cons and is based on a specific security model.

**From my side, I think that any standard protocol for the current scale will work - and I would propose to keep the current implementation except if some planned features mandate a certain protocol selection.**

We have now deployed beszel for around 30 nodes - and this all worked nicely for us. So thanks for the great work 🙂

All the best, and keep up the great work,
Sebastian
PS: in case it would be still a goal to use something like NATS, I can certainly help out there (or anywhere else if time permits 🙃)

**Reply by JSmith-Aura** [reply]
I've talked about this in the discussion above this one, and largely came to the same conclusion you have here. 
And there is now a pull request open for ssh: https://github.com/henrygd/beszel/pull/579

I think NATs might be a good idea later on in the project if we can also work out how to make pocketbase somewhat highly available as using NATs with jetstream for storage might be the go. 
However I think often times perfect is the enemy of good enough and at this scale SSH should be totally fine

**Reply by henrygd** [reply]
Sorry this functionality is taking so long, I've been caught up with work and other things so the next release was delayed a bit.

It should be out fairly soon, then I'll focus on getting this in the next one. I made a [public roadmap](https://github.com/users/henrygd/projects/6) the other day if anyone wants more info about where we stand on something.

I agree that SSH will work fine, or I can adapt JSmith's PR to use HTTP with websockets. I've been thinking the latter might be the easiest from a user perspective as it can be hooked up through an existing reverse proxy or just accessed via an API route instead of a separate port.

@JSmith-Aura Again sorry for the wait, just want to knock out the last things for this release. Please don't worry about conflicts / rebasing. I can take care of that.

**Reply by JSmith-Aura** [reply]
Absolutely no stress at all. Build when you can Im always around to help out if you need it, especially now I've got a bit of familiarity with the code base. 

If you want me to adapt it to using websockets we can actually run ssh over websockets very very easily. Haha

**Comment by freedbygrace** [comment]
After reading up on this protocol, I think this is a good idea and might
have this solution gain even more traction. HenryGD, do you see any
disadvantages?

**Reply by JSmith-Aura** [reply]
There is a PR for this already going.

https://github.com/henrygd/beszel/pull/579
